tests/suite/cli_self_upd.rs

```diff
@@ -304,7 +304,38 @@ info: downloading self-update
             &["rustup", "self", "update"],
             &format!("  rustup updated - {version} (from {version})\n\n",),
             &expected_output,
-        )
+        );
+    });
+}
+
+#[test]
+#[cfg(windows)]
+fn update_overwrites_programs_display_version() {
+    const RUSTUP_UNINSTALL_ENTRY: &str =
+        r"Software\Microsoft\Windows\CurrentVersion\Uninstall\Rustup";
+
+    let version = env!("CARGO_PKG_VERSION");
+
+    update_setup(&|config, _| {
+        config.expect_ok(&["rustup-init", "-y", "--no-modify-path"]);
+
+        winreg::RegKey::predef(winreg::enums::HKEY_CURRENT_USER)
+            .create_subkey(RUSTUP_UNINSTALL_ENTRY)
+            .unwrap()
+            .0
+            .delete_value("DisplayVersion")
+            .unwrap();
```

Comment:
That deletes, not overwrites, thus it tests introducing, not replacing.

I suggest setting the value to a an old release of rustup rather than deleting

Reply:
You are right and I intended to do so, it's just that this test modifies a global state (the winreg). This hasn't been done before, so I'm not sure what is considered proper use of winreg in the tests... Leaving it dirty in an intermediate state seems weird to me.

Reply:
I've chosen `9.999.99`. Hopefully it's clear enough that this is not a real version number.

Reply:
@rami3l oh indeed; we have isolation for such tests already.

Probably this test needs to be reworked - sorry, brain not fully switched on. see https://github.com/rust-lang/rustup/blob/master/src/cli/self_update/test.rs#L38 - thats the core helper, and

https://github.com/search?q=repo%3Arust-lang%2Frustup%20with_saved_path&type=code

which shows it being used in both unit tests and integration tests.

I have no strong view about whether your choice of an integration test or unit is better - though a considerable preference for unit testing rather than integration tests.

Anyhow, what is needed is to extend that helper, or write a similar one, to manage the application uninstall data, and then use that in the test

Reply:
@rbtcollins Regarding this particular case I do think integration test is better since we're ensuring the resolution of that issue (i.e. the correct behavior is exposed to the end user). The current `self update` implementation is a bit convoluted and it took me quite a while to figure out where exactly to add the code modifying `DisplayVersion`...

Reply:
@rbtcollins Done.